pkg/storage/etcd3/watcher.go

```diff
@@ -182,6 +187,11 @@ func (wc *watchChan) startWatching() {
 			wc.sendEvent(parseEvent(e))
 		}
 	}
+	// When we come to this point, it's only possible that client side ends the watch.
+	// e.g. cancel the context, close the client.
+	// If this watch chan is broken and context isn't cancelled, other goroutines will still hang.
+	// We should notify the main thread that this goroutine has exited.
+	close(watchClosedCh)
```

Comment:
Answer is yes.
On error case, we are expecting to return error via the ResultChan() before closing the chan. If we always close it, we won't get any guarantee on that.